packages/query/src/block-processor.ts

```diff
@@ -0,0 +1,390 @@
+      }
+    }
+  }
+
+  private async identifyNewAssets(newNotes: SpendableNoteRecord[]) {
+    for (const n of newNotes) {
+      const assetId = n.note?.value?.assetId;
+      if (!assetId) continue;
+      if (await this.indexedDb.getAssetsMetadata(assetId)) continue;
+
+      let metadata;
+      metadata = await this.querier.shieldedPool.denomMetadata(assetId);
+
+      if (!metadata) {
+        const UNNAMED_ASSET_PREFIX = 'passet';
+        const denom = bech32.encode(UNNAMED_ASSET_PREFIX, bech32.toWords(assetId.inner));
+        metadata = new DenomMetadata({
+          base: denom,
+          denomUnits: [{ aliases: [], denom, exponent: 0 }],
+          display: denom,
+          penumbraAssetId: assetId,
+        });
+      }
+
+      await this.indexedDb.saveAssetsMetadata(metadata);
+    }
+  }
+
+  private async resolveCommitments(commitments?: StoreCommitment[]) {
+    const records = new Map<StateCommitment, SwapRecord | SpendableNoteRecord>();
+
+    // - match commitments to idb notes/swaps
+    for (const storeCommitment of commitments ?? []) {
+      const stateCommitment = StateCommitment.fromJson(storeCommitment.commitment);
+      const record =
+        (await this.indexedDb.getSpendableNoteByCommitment(stateCommitment)) ??
+        (await this.indexedDb.getSwapByCommitment(stateCommitment));
+
+      if (!record) continue;
+
+      records.set(stateCommitment, record);
+    }
+
+    if (commitments?.length) console.log('resolveCommitments', commitments.length, records.size);
+
+    return records;
+  }
+
+  // Nullifier is published in network when a note is spent or swap is claimed.
+  // - if we possess a matching note or swap record, mark it as spent/claimed
+  // - return all records matched
+  private async resolveNullifiers(nullifiers: Nullifier[], height: bigint) {
+    const spentNullifiers = new Map<Nullifier, SpendableNoteRecord | SwapRecord>();
+
+    for (const nf of nullifiers) {
+      const record =
+        (await this.indexedDb.getSpendableNoteByNullifier(nf)) ??
+        (await this.indexedDb.getSwapByNullifier(nf));
+
+      if (!record) continue;
+      console.log('matched record', record);
+
+      spentNullifiers.set(nf, record);
+
+      // record spend/claim
+      if ('heightSpent' in record) {
+        // spendable note
+        record.heightSpent = height;
+        await this.indexedDb.saveSpendableNote(record);
+      } else if ('heightClaimed' in record) {
+        // swap claimed
+        record.heightClaimed = height;
+        await this.indexedDb.saveSwap(record);
+      }
+    }
+
+    if (nullifiers.length)
+      console.log('resolveNullifiers', nullifiers.length, spentNullifiers.size);
+
+    return spentNullifiers;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/require-await
+  async identifyRelevantTransactions(
+    blockSpent: Nullifier[],
+    blockCommitted: StateCommitment[],
+    blockTxs: Transaction[],
```

Comment:
Now there is no problem with transaction detection, but we can also achieve this without giving up rehydrating. 
We pass `blockCommitted: StateCommitment[]`, but we can pass `new_swaps` and `new_notes` instead 

I think the name `identifyRelevantTransactionsAndRehydrate()` would better represent what should be happening in this function 

And handle commitments like this 
```
for (const newNote of newNotes) {
        if (txCommitments.includes(newNote.noteCommitment)) {
          newNote.source = new CommitmentSource({
            source: { case: 'transaction', value: { id: tx } }
          });
          relevantTx.push(tx);
        }
      };

      for (const newSwap of newSwaps) {
        if (txCommitments.includes(newSwap.noteCommitment)) {
          newSwap.source = new CommitmentSource({
            source: { case: 'transaction', value: { id: tx } } }
          });
          relevantTx.push(tx);
        }
      };
```
This is just an illustrative example, I think this will lead to other changes and it could be done better

Reply:
passing new_swaps or new_notes seems incorrect because they have already been written to db, and possibly re-fetched and re-written by nullifier detection step. if we just reuse the object here we might destroy the re-write. and none of this is atomic, so we must fetch again anyway. they are indexed by commitment

the block processor is already complex, so i would like to keep functions as simple as possible and perform a single action. i've written a separate function for rehydration.

Reply:
I don't think we can just leave out the source rehydrate

Reply:
Maybe it makes sense to rewrite the block processor so that notes and swaps are saved only once at the end?

Reply:
It seems to me that we can't detect and spend a note or swap in the same block. So we don't have to worry about `new_notes` or `new_swaps` being overwritten when nullifiers are processed, and then overwritten again when source rehydrate occurs

Reply:
> Maybe it makes sense to rewrite the block processor so that notes and swaps are saved only once at the end?

i was trying to avoid changing the flow in/out of idb, as some of the later steps, some in wasm, reference things written to idb

Reply:
Yes, I agree, the in/out of idb flow really does not look transparent because of wasm